doc: specify how caching CMM should handle max plaintext length in Get Encryption Materials (#159)
awslabs/aws-encryption-sdk-specification: changes/2020-07-13_caching-cmm-max-plaintext-length/change.md

## Affected Implementations

| Language   | Repository                                                                            |
| ---------- | ------------------------------------------------------------------------------------- |
| C          | [aws-encryption-sdk-c](https://github.com/aws/aws-encryption-sdk-c)                   |
| Java       | [aws-encryption-sdk-java](https://github.com/aws/aws-encryption-sdk-java)             |
| JavaScript | [aws-encryption-sdk-javascript](https://github.com/aws/aws-encryption-sdk-javascript) |
| Python     | [aws-encryption-sdk-python](https://github.com/aws/aws-encryption-sdk-python)         |

## Definitions

### Conventions used in this document

The key words
"MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT",
"SHOULD", "SHOULD NOT", "RECOMMENDED", "MAY", and "OPTIONAL"
in this document are to be interpreted as described in
[RFC 2119](https://tools.ietf.org/html/rfc2119).

## Summary

The Get Encryption Materials operation accepts an optional `max plaintext length` parameter,
but the specification does not state how the caching CMM should behave
when the caller does not provide the parameter value.

> This change specifies that the caching CMM should bypass the cache in this case.

Author: Thanks, will fix this.

Also, when the caching CMM is performing a Get Encryption Materials operation
for which no materials are cached,
it MUST call its underlying CMM's Get Encryption Materials operation.
The specification does not state what value of `max plaintext length` (if any)
the caching CMM should pass to its underlying CMM.

> This change specifies that the caching CMM should pass its `byte limit` value

Author: Thanks, will fix this.

as the `max plaintext length` parameter of the call to the underlying CMM.

## Out of Scope

- Changing the shape of any CMM APIs is out of scope.

## Motivation

### Fulfulling Encryption Materials Requests Without `max plaintext length`

The Get Encryption Materials operation of the Cryptographic Materials Manager (CMM)

> accepts an Encryption Materials Request that MAY specify a `max plaintext length` field,

Reviewer: Good catch. We should have this be "MUST optionally accept a `max plaintext length`" or similar. We use MAY somewhat incorrectly in our spec.

What we want is that the implementation MUST accept this value if it's provided, but it's optional for the customer to provide it.

indicating the maximum length of plaintext
that the caller intends to encrypt using the returned materials.

> The caching Cryptographic Materials Manager (caching CMM)

Reviewer: np: You should probably move `Cryptographic Materials Manager (caching CMM)` higher up (to the first time you use `caching CMM`) instead of defining it after it's been used a few times.

Author: Thanks, I'll fix this.

aims to prevent its users from using encryption materials
to encrypt more bytes than its configured `byte limit`,
and existing implementations do so by bypassing the cache
whenever it does not know their users' `max plaintext length`.
This behavior is safe but previously unspecified,
and this change mandates that implementations implement it.

### Setting `max plaintext length` for Underlying CMM's Get Encryption Materials

The CMM interface specification implicitly allows an underlying CMM
to produce encryption materials using logic

> that depends on the `max plaintext length` of its Encryption Materials Request.

Author: I agree that "its" isn't right here, but I think substituting "an" makes more sense than dropping it here.

Suppose the caching CMM passes a `max plaintext length` value of `N`
to its underlying CMM's Get Encryption Materials operation,
and caches the materials that it receives.
If a later Get Encryption Materials operation fetches those materials from the cache,
but has a `max plaintext length` value that exceeds `N`,

> then the caching CMM returns those materials unsafely.

Reviewer 2: What do we mean by "unsafely"? We should say exactly what we mean here. By passing `N` to the underlying CMM, this CMM is signaling to the underlying CMM that it intends to only use the returned materials to encrypt up to `N` bytes. By then reusing those materials such that when it reuses the materials it encrypts plaintext with length exceeding `N`, this CMM is violating it's previous intent. This violation is why this supposed approach is bad.

Author: @lavaleri thanks, I've incorporated your wording to clarify the explanation.

After a discussion with @lavaleri, @seebees, and @mattsb42-aws offline, we think that the edge case that @seebees described requires more thought to resolve properly, and doing so may require changes to other parts of this PR. So we are moving the PR to a later milestone.

It is therefore important that the caching CMM
carefully selects the `max plaintext length` parameter to pass to its underlying CMM.
This change mandates that the caching CMM passes its configured `byte limit` value
as the `max plaintext length` parameter to the underlying CMM,
whereas previously, the caching CMM could choose any value.
With this change in place,
even if the underlying CMM produces materials using plaintext-length-dependent logic,
the caching CMM receives and caches materials
that are safe to use for up to `byte limit` bytes of plaintext.
It follows that if a later Get Encryption Materials operation's `max plaintext length`
does not exceed the caching CMM's `byte limit`,

> then it is also safe for the caching CMM to return corresponding cached materials.

Reviewer 2: We should mention that the cache keeps a running total of the max bytes that the user indicated they will encrypt using the materials. Otherwise, it is unclear why this is true.

Author: Thanks, I will clarify this.

## Drawbacks

Bypassing the cache in the caching CMM on Get Encryption Materials operations
that do not specify `max plaintext length`
could be confusing to the user.
Warning the user when this happens could help prevent users from running into
unexpected (and potentially expensive) calls to the underlying CMM,
but we leave this as an implementation choice.

## Security Implications

This change SHOULD NOT have any security implications.

## Operational Implications

This change will break any customer use case
that depends on the Caching CMM passing no `max plaintext length` parameter
to the underlying CMM on Get Encryption Materials calls.

## Guide-level Explanation

> When you make a Get Encryption Materials call to the caching CMM,

Reviewer 2: here and throughout, I think we should frame this in terms of "a caller" instead of "you". A caller could be the encrypt operation, or it could be another CMM.

you may optionally specify a value for the `max plaintext length` parameter,
which indicates the maximum number of bytes you will encrypt
with the materials you receive from the call.
If you do not specify a `max plaintext length` value,
then the caching CMM will neither use cached materials
nor cache materials from its underlying CMM.
This helps to ensure that you do not exceed the `byte limit`
that you configure on the caching CMM.
So in order to maximize the benefits of caching,
you should specify the `max plaintext length` value

> whenever you know an appropriate upper bound for bytes you will encrypt.

Reviewer 3: nit: it is true that this will _maximize the benefits_... but the overriding reason is that it is **not safe** to encrypt and unlimited number of bytes under a single data key.

Author: Thanks, I will change the wording to clarify this.

It follows that if a later Get Encryption Materials operation's `max plaintext length`
does not exceed the remaining number of bytes for which the cached materials can be safely used
(i.e.,
the difference between the caching CMM's `byte limit`
and the sum of `max plaintext length` values from past uses of the cached materials),
then it is also safe for the caching CMM to return the cached materials.

## Drawbacks

Bypassing the cache in the caching CMM on Get Encryption Materials operations
that do not specify `max plaintext length`
could be confusing to the user.
Warning the user when this happens could help prevent users from running into
unexpected (and potentially expensive) calls to the underlying CMM,
but we leave this as an implementation choice.

## Security Implications

This change SHOULD NOT have any security implications.

## Operational Implications

This change will break any customer use case
that depends on the Caching CMM passing no `max plaintext length` parameter
to the underlying CMM on Get Encryption Materials calls.

## Guide-level Explanation

When making a Get Encryption Materials call to the caching CMM,

> the user may optionally specify a value for the `max plaintext length` parameter,

Reviewer 4: ```suggestion
the user MAY optionally specify a value for the `max plaintext length` parameter,
```

which indicates the maximum number of bytes the user intends to encrypt
with the materials received from the call.
If the user does not specify a `max plaintext length` value,
then the caching CMM will neither use cached materials

> nor cache materials from its underlying CMM.

Reviewer 4: I think we can simplify this to say something along the lines of "the caching CMM will pass through to the underlying CMM without involving the cache".

This helps to ensure that the number of bytes encrypted with the materials
does not exceed the `byte limit` that the user configures on the caching CMM,
while also maximizing the benefits of caching.

> So whenever the user can determine an appropriate upper bound for bytes they will encrypt,
> they should specify the `max plaintext length` value.

Reviewer 4: ```suggestion
If the user can determine an appropriate upper bound for bytes they will encrypt,
they SHOULD specify the `max plaintext length` value.
```

If the user customizes the underlying CMM of a caching CMM,
then whenever the caching CMM makes a Get Encryption Materials call
to the customized underlying CMM,
the Encryption Materials Request will have a `max plaintext length` value

> equal to the `byte limit` value configured on the caching CMM.

Reviewer 4: What does "customizes the underlying CMM" mean in this case? This statement is true regardless of what underlying CMM you use.
